app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt

```diff
@@ -0,0 +1,421 @@
+import com.fsck.k9.fragment.MLFProjectionInfo.FLAGGED_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.FOLDER_SERVER_ID_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.FORWARDED_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.PREVIEW_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.PREVIEW_TYPE_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.READ_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.SENDER_LIST_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.SUBJECT_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.THREAD_COUNT_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.TO_LIST_COLUMN
+import com.fsck.k9.fragment.MLFProjectionInfo.UID_COLUMN
+
+import kotlin.math.max
+
+class MessageListAdapter internal constructor(
+        context: Context,
+        theme: Resources.Theme,
+        private val res: Resources,
+        private val layoutInflater: LayoutInflater,
+        private val messageHelper: MessageHelper,
+        private val contactsPictureLoader: ContactPictureLoader,
+        private val preferences: Preferences,
+        private val listItemListener: MessageListItemActionListener,
+        private val showingThreadedList: Boolean = false
+) : CursorAdapter(context, null, 0) {
+
+    private val forwardedIcon: Drawable
+    private val answeredIcon: Drawable
+    private val forwardedAnsweredIcon: Drawable
+    private val previewTextColor: Int
+    private val activeItemBackgroundColor: Int
+    private val selectedItemBackgroundColor: Int
+    private val readItemBackgroundColor: Int
+    private val unreadItemBackgroundColor: Int
+    private val fontSizes = K9.fontSizes
+
+    init {
+
+        val attributes = intArrayOf(
+                R.attr.messageListAnswered,
+                R.attr.messageListForwarded,
+                R.attr.messageListAnsweredForwarded,
+                R.attr.messageListPreviewTextColor,
+                R.attr.messageListActiveItemBackgroundColor,
+                R.attr.messageListSelectedBackgroundColor,
+                R.attr.messageListReadItemBackgroundColor,
+                R.attr.messageListUnreadItemBackgroundColor
+        )
+
+        val array = theme.obtainStyledAttributes(attributes)
+
+        answeredIcon = res.getDrawable(array.getResourceId(0, R.drawable.ic_messagelist_answered_dark))
+        forwardedIcon = res.getDrawable(array.getResourceId(1, R.drawable.ic_messagelist_forwarded_dark))
+        forwardedAnsweredIcon = res.getDrawable(array.getResourceId(2, R.drawable.ic_messagelist_answered_forwarded_dark))
+        previewTextColor = array.getColor(3, Color.BLACK)
+        activeItemBackgroundColor = array.getColor(4, Color.BLACK)
+        selectedItemBackgroundColor = array.getColor(5, Color.BLACK)
+        readItemBackgroundColor = array.getColor(6, Color.BLACK)
+        unreadItemBackgroundColor = array.getColor(7, Color.BLACK)
+
+        array.recycle()
+    }
+
+    private inline val previewLines: Int
```

Comment:
Missing the MessageListAppearance data class.

Reply:
Looks like this change didn't make it into #4109. I suggest we add the class in a separate pull request after this one has been merged.

Reply:
Sure.